bitclock-fw/main/libs/mqtt.c

```diff
@@ -0,0 +1,123 @@
+#include "mqtt.h"
+#include "esp_log.h"
+#include "mqtt_client.h"
+
+static const char *TAG = "mqtt";
+
+esp_mqtt_client_handle_t client;
+static bool configs_published = false;
+
+static const char *co2_unique_id = "bitclock_co2";
+static const char *nox_unique_id = "bitclock_nox";
+static const char *voc_unique_id = "bitclock_voc";
+static const char *temp_unique_id = "bitclock_temp";
+static const char *humidity_unique_id = "bitclock_humidity";
+static char config_message[300];
+static char config_topic[100];
+static char state_topic[100];
+
+bool publish_message(const char *topic, const char *message, bool retain) {
+  bool success =
+      esp_mqtt_client_publish(client, topic, message, 0, 1, retain) >= 0;
+  if (success) {
+    ESP_LOGI(TAG, "MQTT publish successful, topic=%s message=%s", topic,
+             message);
+  } else {
+    ESP_LOGE(TAG, "MQTT publish failed, topic=%s message=%s", topic, message);
+  }
+  return success;
+}
+
+bool publish_homeassistant_config(const char *name, const char *unique_id,
+                                  const char *device_class,
+                                  const char *unit_of_measurement,
+                                  const char *state_class,
+                                  const char *platform) {
+  snprintf(state_topic, sizeof(state_topic), "homeassistant/sensor/%s/state",
+           unique_id);
+  snprintf(config_message, sizeof(config_message),
+           "{\"name\": \"%s\",\"uniq_id\": \"%s\", \"dev_cla\": \"%s\", "
+           "\"unit_of_meas\": \"%s\",\"stat_cla\": "
+           "\"%s\",\"p\": \"%s\", \"stat_t\": \"%s\"}",
+           name, unique_id, device_class, unit_of_measurement, state_class,
+           platform, state_topic);
```

Comment:
@glenrobertson I was getting errors if I didn't supply `state_topic` so I added it. Also used abbreviations to shorten the json.